### ISSUE TITLE
parallelisation is not working in optuna

### ISSUE THREAD

**Issue by ali3assi** [optuna/optuna#2351]
How can we execute the code in parallel in optuna. The documentation is so naive about that.
Suppose I have a test.py file which includes the following code:

```
 study = optuna.create_study(
        direction="maximize",
        study_name="ali", #for parallel execution you need these couple of lines 
        storage="sqlite:///mydb.db", 
        load_if_exists= True, #to skip creating a new study if it already exists. 
        pruner = SuccessiveHalvingPruner(),
        #pruner=HyperbandPruner(min_resource=1, max_resource=100, reduction_factor=3),
    )
```

and I have another test2.py file which calls the test.py file like that:
```
import multiprocessing as mp
from concurrent.futures import ProcessPoolExecutor
import os
from multiprocessing import Pool

def run_process (process):
    os.system('python {}'.format(process))

if __name__ == "__main__":
    processes = ['TestDriver.py']
    pool = Pool()
    pool.map(run_process, processes)
    pool.close    
```
  But the code:

1. is not executed in parallel. I cannot see for example in the console: Trial 0 finsihed .... Trial 1....
2. The code creates two experiments in neptune.

## Note to the questioner

If you are more comfortable with [Stack Overflow], you may consider posting your question with the "optuna" tag there instead.
Alternatively, for issues that would benefit from more of an interactive session with the developers,
you may refer to the [optuna/optuna chat] on Gitter.

[Stack Overflow]: https://stackoverflow.com/questions/tagged/optuna
[optuna/optuna chat]: https://gitter.im/optuna/optuna

**Comment by HideakiImamura** [comment]
Thanks for the question. Please execute the python program in different processes manually. For example, if you use macOS, open 2 terminals, and just run `python test2.py` in each terminal. Then, Optuna parallelizes the optimization with 2 processes.

**Comment by ali3assi** [comment]
@HideakiImamura  Thanks for your reply. I am on windows. But I need to parallelize the code on a pc of 48 cpus, I need to open 48 terminales?

**Comment by HideakiImamura** [comment]
Actually, you don't have to open 48 windows. You can just open one terminal and execute the python program twice in the background.

**Comment by ali3assi** [comment]
@HideakiImamura Sorry sir but it is not clear. How do that in the background?

**Comment by HideakiImamura** [comment]
Just run `python test2.py &`.

**Comment by ali3assi** [comment]
Thanks but python test2.py & how many times this command suppose to execute the code? It is equivalent to how many "opened' terminal?

**Comment by HideakiImamura** [comment]
Once. This command will simply run `test2.py` once in the background.

**Comment by ali3assi** [comment]
@HideakiImamura I am so sorry sir, I cannot catch your idea. I need to parallelize. I just Try the command and the code is executed in sequential way for one time. So suppose i do this command how it is executed 48 times

**Comment by ali3assi** [comment]
I any way is not working in parallel, two experiments are created while logically i should have only one which runs in parallel

**Comment by HideakiImamura** [comment]
You seem to have misunderstood Optuna's parallel distributed optimization scheme. Optuna's parallelization scheme is how to run trials in parallel. The study, which is equal to the experiment, is identified with the name. If you use `load_if_exists=True`, the existing study is loaded. In different processes, `study.optimize`s communicate with each other through your specified RDB. 

If you want to parallelize your objective function, please implement it so. Of course, this cannot be controlled by Optuna.

**Comment by koh-joshua** [comment]
I think what ali3assi may be asking for is the good old n_jobs=-1:
study.optimize(objective, n_trials=n_trials, n_jobs=-1)

This will utilize all the cores on your computer for the single study. However, depending on the complexity of the objective and number of trials, you will typically not see all cores utilized or 100% CPU usage. Also, since Optuna > 2.0, the n_jobs=-1 don't really seem to be efficient in using multiple cores. Up until version 2.0, I can set n_jobs=-1 and launch multiple instances of Optuna and see my **CPU usage hit 100% for 32 cores sustained**. After 2.0, it seems that each study is locked to a process or there's some limitations and this prevents my CPU usage scaling to 100%. The speed penalty is extreme. 

For now, version 2.0 is gold to me.

What about distributed parallelism in Optuna using RDB? That is ok if your trial run is up to 100,000 or less, split or distributed between multiple instances of Optuna for the same study. It gets extremely slow after that. But due to the random seed initialization in the optimizers, you'd be better running multiple instances of Optuna to get a better tuned model or objective. At least in my own experience, I recover the best result from 5 instances/5 studies of 20,00 trials each versus one instance/study at 100,000 trials.

Hope this helps.

**Comment by not522** [comment]
@koh-joshua Thank you. Your observation is very meaningful. Could you give a reproducible example? It will help to investigate the cause of the performance problem.

**Comment by nzw0301** [comment]
I could reproduce the reported issue by @koh-joshua with a single study setting:

```python
import sklearn.datasets
import sklearn.linear_model
import sklearn.model_selection

import optuna

optuna.logging.set_verbosity(optuna.logging.WARN)

def objective(trial):
    iris = sklearn.datasets.load_iris()
    classes = list(set(iris.target))
    train_x, valid_x, train_y, valid_y = sklearn.model_selection.train_test_split(
        iris.data, iris.target, test_size=0.25
    )

    alpha = trial.suggest_float("alpha", 1e-5, 1e-1, log=True)
    clf = sklearn.linear_model.SGDClassifier(alpha=alpha)

    for step in range(100):
        clf.partial_fit(train_x, train_y, classes=classes)

    return clf.score(valid_x, valid_y)

storage = optuna.storages.RDBStorage(url="mysql://root@localhost/example")

# n_job=-1
%timeit -r 3 -n 1 optuna.create_study(direction="maximize", storage=storage).optimize(objective, n_trials=36, n_jobs=-1)
# 6.6 s ± 420 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)

# n_job=1
%timeit -r 3 -n 1 optuna.create_study(direction="maximize", storage=storage).optimize(objective, n_trials=36)
# 6.11 s ± 136 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)
```

Note that the code block comes from jupyter notebook to measure the running time.

Even though I checked out the optuna's version to `v1.5.0`, `n_job=-1`'s performance was not faster than `n_job=1`'s one.

To make optuna faster again, I tried to change the backend of `joblib` of optuna `v.2.4.0` explicitly; 

I replaced [this line of `optimize` method](https://github.com/optuna/optuna/blob/v2.4.0/optuna/_optimize.py#L94) with 
```
with Parallel(n_jobs=n_jobs, prefer="parallel") as parallel:
```

Then,  🚀 

```python
%timeit -r 3 -n 1 optuna.create_study(direction="maximize", storage=storage).optimize(objective, n_trials=36, n_jobs=-1)
3.75 s ± 2.71 s per loop (mean ± std. dev. of 3 runs, 1 loop each)

%timeit -r 3 -n 1 optuna.create_study(direction="maximize", storage=storage).optimize(objective, n_trials=36)
5.86 s ± 258 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)
```

I'm sorry if my approach above is wrong and confuses you.

---

Environments:
  - #cores: 6
  - OS: Mac OS Big Sur 11.2.1
  - optuna: b6de94d1f3042b43d5a1cb5b729795d74cebb026
  - MySQL: 8.0.23

**Comment by codeananda** [comment]
For those not running Windows, you could use [terminator](https://gnometerminator.blogspot.com/p/introduction.html) to easily create a tile of N terminals and use the 'Broadcast All' option to type in and execute the same code in all terminals at once. 

<img width="501" alt="image" src="https://user-images.githubusercontent.com/51246969/201105596-c5986299-92e6-4a7a-8aad-da6b89f43fb6.png">

There are other answers and options available in this [stackexchange question](https://unix.stackexchange.com/questions/89339/how-do-i-run-the-same-linux-command-in-more-than-one-tab-shell-simultaneously). I adapted [this answer](https://unix.stackexchange.com/a/89471).
